server/src/main/java/org/cloudfoundry/identity/uaa/impl/config/AwsSecretsLoader.java

```diff
@@ -0,0 +1,86 @@
+
+public class AwsSecretsLoader {
+
+    private final ObjectMapper mapper = new ObjectMapper();
+    private static final Logger LOGGER = LoggerFactory.getLogger(AwsSecretsLoader.class);
+
+    public void loadSecrets(ConfigurableEnvironment environment) {
+        String awsSecretNames = environment.getProperty("AWS_SECRET_MANAGER_NAMES");
+        if (StringUtils.hasText(awsSecretNames)) {
+            String[] secretNames = awsSecretNames.split(";");
+            AWSSecretsManager secretManagerClient = awsSecretsManager();
+            Arrays.stream(secretNames).forEach(secretName -> {
+                try {
+                    LOGGER.info("Loading secrets from AWS Secret Manager for Secret Name {}", secretName);
+
+                    GetSecretValueResult results =
+                            secretManagerClient.getSecretValue(new GetSecretValueRequest().withSecretId(secretName));
+                    JsonNode jsonNode = mapper.readTree(results.getSecretString());
+                    Map<String, Object> map = mapper.convertValue(jsonNode, Map.class);
+                    Map<String, Object> finalMap = new LinkedHashMap<>();
+                    map.forEach((key, value) -> {
+                        if (value instanceof Map) {
+                            Map<String, Object> valFromEnv = environment.getProperty(key, Map.class);
+                            if (valFromEnv != null) {
+                                deepMergeMaps((Map<String, Object>) value, valFromEnv);
+                            } else {
+                                finalMap.put(key, value);
+                            }
+                        } else {
+                            finalMap.put(key, value);
+                        }
+                    });
+
+                    String propertySourceName = "aws.secret.manager" + secretName.replace("/", ".");
+                    environment.getPropertySources().addLast(new MapPropertySource(propertySourceName, finalMap));
+                } catch (Exception e) {
+                    LOGGER.error("Unable to load secrets from AWS Secret Manager for Secret Name {}", secretName, e);
+                }
+            });
+        } else {
+            LOGGER.info("AWS Secret Manager is not configured");
+        }
+    }
+
+    protected AWSSecretsManager awsSecretsManager() {
+        return AWSSecretsManagerClientBuilder.standard().withClientConfiguration(
+                SpringCloudClientConfiguration.getClientConfiguration()).build();
+    }
+
+    private Map<String, Object> deepMergeMaps(Map<String, Object> source, Map<String, Object> target) {
+        for (String key : source.keySet()) {
+            Object value = source.get(key);
+            if (!target.containsKey(key)) {
+                // new value for "key"
+                target.put(key, value);
+            } else {
+                // existing value for "key" - recursively deep merge
+                if (value instanceof Map) {
```

Comment:
Line 78, you are assuming target map value also will be a map, better add one more condition like below,

`if (value instanceof Map && target.get(key) instanceof Map) {`

Reply:
Done